File: library/swift/src/stats/HistogramTimerImpl.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// The implementation of histogram specific to tracking time durations
@objcMembers
class HistogramTimerImpl: NSObject, Histogram {
  private let series: String
  private weak var engine: EnvoyEngine?

  init(elements: [Element], engine: EnvoyEngine) {
    self.series = elements.map { $0.value }.joined(separator: ".")
    self.engine = engine
    super.init()
  }

  /// Record a new value for the histogram distribution.
  /// TODO: potentially raise error to platform if the operation is not successful.
  func recordValue(value: Int) {
    guard let engine = self.engine else {
      return
    }

    engine.recordHistogramValue(self.series, value: numericCast(value), unitMeasure: MICROSECONDS)
  }
}

File: library/kotlin/src/io/envoyproxy/envoymobile/stats/HistogramTimerImpl.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyEngine
import io.envoyproxy.envoymobile.engine.types.HistogramUnit
import java.lang.ref.WeakReference

/**
 * Envoy implementation of a `Histogram` for time measurements e.g. durations.
 */
internal class HistogramTimerImpl : Histogram {
  internal val envoyEngine: WeakReference<EnvoyEngine>
  internal val elements: List<Element>

  internal constructor(engine: EnvoyEngine, elements: List<Element>) {
    this.envoyEngine = WeakReference<EnvoyEngine>(engine)
    this.elements = elements
  }

  override fun recordValue(value: Int) {
    envoyEngine.get()?.recordHistogramValue(
      elements.joinToString(separator = ".") { it.value }, value, HistogramUnit.MICROSECONDS
    )
  }
}
